src/sentry/migrations/0634_backfill_github_webhook_outbox_shard_ids.py

```diff
@@ -0,0 +1,85 @@
+# Generated by Django 3.2.23 on 2024-01-19 05:45
+
+from django.db import migrations
+
+from sentry.new_migrations.migrations import CheckedMigration
+from sentry.utils import json
+from sentry.utils.query import RangeQuerySetWrapperWithProgressBar
+from sentry.utils.types import Any, Dict
+
+
+def validate_payload(outbox) -> bool:
+    payload = outbox.payload
+    body: str = payload.get("body", None)
+    if not payload or not body:
+        return False
+
+    headers = payload.get("headers", None)
+    if not headers or "application/json" not in headers.get("Content-Type", ""):
+        return False
+
+    return True
+
+
+def get_integration_id_from_payload(parsed_body, IntegrationModel) -> int | None:
+    installation_id = parsed_body.get("installation", {}).get("id", None)
+    if not installation_id:
+        return None
+
+    integration = IntegrationModel.objects.filter(
+        external_id=installation_id, provider="github"
+    ).first()
+
+    return integration.id if integration else None
+
+
+def backfill_webhook_outbox_shard_ids(apps, schema_editor):
+    ControlOutbox = apps.get_model("sentry", "ControlOutbox")
+    Integration = apps.get_model("sentry", "Integration")
+
+    for outbox in RangeQuerySetWrapperWithProgressBar(
+        ControlOutbox.objects.filter(
+            category=1,
+            shard_scope=2,  # Webhook proxy category  # Webhook scope
+            shard_identifier=1,  # GitHub Provider Identity
+        )
+    ):
+        if not validate_payload(outbox=outbox):
+            continue
+
+        body: Dict[str, Any] = json.loads(outbox.payload.get("body"))
+
+        integration_id = get_integration_id_from_payload(
+            parsed_body=body, IntegrationModel=Integration
+        )
+        if integration_id is None:
+            continue
+
+        outbox.shard_identifier = integration_id
+        outbox.save()
```

Comment:
We *could* hit contention/locks here as the outbox delivery task uses row locks.
```suggestion
        try:
            outbox.save()
        except Exception:
             pass
```

We could catch any errors and continue moving forward as skipping locked rows is better than getting stuck.